### PR TITLE
#188 plugins.legend

### PR DESCRIPTION
**Pull request by Ytawo**
#188

### CODE REVIEW

**Comment by sheymann** [comment]
Bug on first load of the example page:

> Uncaught InvalidStateError: Failed to execute 'drawImage' on 'CanvasRenderingContext2D': The HTMLImageElement provided is in the 'broken' state.

Way to reproduce:
1. open plugin-legend.html in a Chrome private tab
- open the console

image

This exception disappear after reloading the page, so it is probably due to the font not being loaded yet.

**Comment by sheymann** [comment]
- [ ] need license, we'll use GPL like the design plugin https://github.com/Linkurious/linkurious.js/tree/develop/plugins/sigma.plugins.design

**Comment by sheymann** [comment]
- [ ] styling: can we set the widget titles left-aligned instead of centered?

**Comment by sheymann** [comment]
- why a yellow color for shapes and sizes in the example? A grey color would be better because it is neutral: shapes and sizes and not encoded by color

**Comment by Ytawo** [comment]
@sheymann It is configurable. I set orange just because I wanted to test with a color other than black. It is the "legendShapeColor" setting.

**Comment by sheymann** [comment]
- [ ] no need for brackets in text "(source node to target node)", we understand that it is a subtitle
- [ ] truncate the property key in title if title length > threshold (better to read the first characters than not reading because the font is too small)

**Comment by sheymann** [comment]
The whole legend disappears if not enough space to display all widgets. Is it a feature or a bug? Please edit the README if it is a feature.

**Comment by Ytawo** [comment]
@sheymann It's a feature! I'll add it to the README.

**Comment by sheymann** [comment]
@Ytawo okay let's set default shape color to black. The only good reason to change it is to invert color of background/font to black/white, or use a lighter black.

**Comment by sheymann** [comment]
- [ ] missing legendBorderRadius setting :)

**Comment by sheymann** [comment]
What if the same property is mapped to both color and size? What can we do to reduce the space occupied by the legend?

**Comment by sheymann** [comment]
- [ ] Icons and texts are slightly misaligned, with icons above the text baseline:

image

**Comment by sheymann** [comment]
- [ ] code style: can to use lines with max length 120 chars please? It will be easier to read.

**Comment by sheymann** [comment]
- [ ] code style: I suggest to move all private function to top of the public ones. It is the case in all other plugins.

**Comment by sheymann** [comment]
- [ ] README: please add link to the example file, see the beginning of https://github.com/Linkurious/linkurious.js/tree/develop/plugins/sigma.plugins.locate

**Comment by sheymann** [comment]
How is the order of widgets defined? I'd like to display all node related widgets before edge related widgets.

**Comment by sheymann** [comment]
- [ ] typo in README: "sigma.utils.legend" instead of "sigma.plugins.legend", same for killLegend

**Comment by sheymann** [comment]
- [ ] truncate property keys as well (need a second truncate param?)

image

Would it be a way to avoid using another setting by computing the longest string possible? Is there an equivalent in svg to `canvas.measureText()`?

**Comment by sheymann** [comment]
- [ ] in example, edge shape values should be qualitative instead of quantitative because shape cannot visually encode quantities. Can you find another set of values to represent edge categories as shapes please?

**Comment by sheymann** [comment]
- [ ] How to display "long" numbers? (could be large values or floats)

image

image

**Comment by Ytawo** [comment]
@sheymann For numbers, I suggest that we fix their size to 4 characters:

* if > 9999: 3 most significant digit + unit (K, M, G, ...)
* if <= 9999: 4 most significant digits (including digits after the comma)
* if the text is still too long, reduce the font size

Moreover, assuming that you're ok with that method, should we display large numbers without shrinking them if there is enough space? (I personally think that we should shrink them in either case, most people would probably prefer read "172M" instead of "172093713.83", even if there is enough space).

**Comment by sheymann** [comment]
- [ ] Legend widgets should be horizontally aligned by their bottom:

image

**Comment by sheymann** [comment]
@Ytawo I fully agree about your suggestions

**Comment by sheymann** [comment]
- [ ] font of icons not loaded when creating the legend:

image

Force the fond to load by adding this in the example code:

```html
<!-- we force the loading of the font awesome -->
<i class="fa fa-camera-retro fa-1x" style="color: rgba(0,0,0,0);"></i>
```

NB: font is loaded only if bound to a style of an existing DOM element.

Combine it with the following code to init sigma after the font is loaded:

```js
// In case the WebGL renderer is used, we must wait for FontAwesome to be loaded.
// http://www.w3.org/TR/css-font-loading/
if (document.fonts) {
  // document.fonts.ready() method is going to be replaced with
  // document.fonts.ready attribute in the future.
  var fontsReady = document.fonts.ready;
  if (typeof(fontsReady) == "function") {
    fontsReady = document.fonts.ready();
  }
  fontsReady.then(init);
}
else {
  // wait or use a polyfill such as:
  // https://github.com/zachleat/fontfaceonload
  // https://github.com/smnh/FontLoader
  setTimeout(init, 2000);
}
```

ref: renderers-linkurious-icons.html

**Comment by sheymann** [comment]
- [ ] "source node to target node" should be left aligned

**Comment by sheymann** [comment]
What if a mapping for a property value is described in plugins.designed but the property doesn't exist in the graph? E.g. in the example, what if the "Man" value of property "Category" is described but no node has the "Man" "Category"?

Is it feasible to detect it when generating the legend?

**Comment by sheymann** [comment]
- [ ] the layout is weird on this image, ideally columns heights would not have large differences:

image

Text and Age could be on the same column, as well as Family name and Relation Strength

- [ ] A different issue below, the width is not fully used, resulting in unnecessary overlapping of the viz by the legend:
image

**Comment by sheymann** [comment]
- [ ] bug on right placement, the position of the text widget is not stable with slight change of window height:

image

vs

image

**Comment by sheymann** [comment]
- [ ] Please write comments on public methods of the legend plugin such as `LegendPlugin.prototype.setExternalCSS`
- [ ] Please also write comments on internal functions to ease maintenance of the plugin
- What is the difference between `drawSizeLegend` and `drawNonSizeLegend`?

**Comment by sheymann** [comment]
- [ ] documentation is missing in README about widgets (e.g. how to pin widgets)

**Comment by Ytawo** [comment]
* The widget.setPosition pins widgets (it is described in the README)
* The setExternalCSS is useless right now, we will probably need it for the export
* yes it is feasable to detect if a category is not used within a graph, I will add it

For the layout, I tried to do as close as possible as what we described. I will equalize the size of the columns, but for the text widget being alone on a column, I thought that was what you meant by "blocks".

**Comment by Ytawo** [comment]
I have reworked the layout algorithm, columns should be more equal and the widgets' position should not change too much when a resizing happen.

**Comment by sheymann** [comment]
Much better layout!

- [ ] In README I suggest to separate more the Legend API from the LegendWidget API, currently it's too confusing to read.

- [ ] In the comments of the source code, the type of function parameters are always missing (e.g. line 587, `@param unit` should be `@param {string} unit`;

- [ ] Multiple comments on methods are missing, e.g. the method `LegendPlugin.prototype.drawLayout` has no comment to describe what the method does and what its side effets are.

- [ ] Multiple methods such as `LegendPlugin.prototype.drawLayout` are public but they are not described in README. It seems that these methods should be private.

File: plugins/sigma.plugins.legend/README.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+ * @param visualVar   {string} 'color', 'icon', 'size', 'type'
+ * @returns {LegendWidget}
+ */
+legend.getWidget(elementType, visualVar);
+
+/**
+ * Build the widgets and redraw the legend.
+ * Must be called whenever the graph's design changes
+ */
+legend.redraw();
```
Comment (sheymann) [inline plugins/sigma.plugins.legend/README.md]: How about `draw` instead of `redraw?`

**Comment by Ytawo** [comment]
Initially, "draw" = only draw, and "redraw" = build widgets + compute layout + draw.
But I agree, it's confusing, I'll change it to "draw".

**Pull request merged**